README.md
Elisa Viihde Python library
=====

This library is not maintained by Elisa so read it unofficial.

_Elisa Viihde is a registered trademark of [©Elisa Oyj](http://corporate.elisa.fi)_

**License:** GPLv3 http://www.gnu.org/copyleft/gpl.html

**Requires:** Requests http://docs.python-requests.org/
* **Very important fact:** By default, urllib3 (which Requests uses) does not verify HTTPS requests. https://urllib3.readthedocs.org/en/latest/security.html

**Developed for:** Python 2.7.6

[![Build Status](https://travis-ci.org/enyone/elisaviihde.svg?branch=master)](https://travis-ci.org/enyone/elisaviihde)
[![Coverage Status](https://coveralls.io/repos/enyone/elisaviihde/badge.svg?branch=master)](https://coveralls.io/r/enyone/elisaviihde?branch=master)
[![Gitter](https://badges.gitter.im/Join%20Chat.svg)](https://gitter.im/enyone/elisaviihde?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge)

**Latest release package**

[Download 1.4 zip](https://github.com/enyone/elisaviihde/archive/1.4.zip)
```
sha256sum: 1257c0190b735ecac61aea9b1cd630d2c6611a3e262bbc7df7a0281732f76e42
```

[Download 1.4 tar.gz](https://github.com/enyone/elisaviihde/archive/1.4.tar.gz)
```
sha256sum: b33f6b0b215f7dbfb02bc91dad6c730d3c255ebb966c839225d4071c0c3c4457
```

**KODI/XBMC Python module addon**

[Download script.module.elisaviihde-1.4.0.zip](https://github.com/enyone/elisaviihde/releases/download/1.4/script.module.elisaviihde-1.4.0.zip)
```
sha256sum: 85ba43945be12a8da76dd4b0f6be49700c70ee34b5e654f825d87efec4fdbd73
```

API documentation
-----

https://github.com/enyone/elisaviihde/wiki

Simple example
-----
```
$ cp examples/example.py .
$ python example.py -u username
Found folders:
3603265: Ajankohtainen kakkonen
2540806: Dokumentit
2540838: Elokuvat

Found recordings from folder 3603265:
1812084: Ajankohtainen kakkonen (ke 25.02.2015 13.25)
1811241: Ajankohtainen kakkonen (ti 24.02.2015 21.00)
1797570: Ajankohtainen kakkonen (ke 18.02.2015 14.45)

Found stream uri from recording 1812084:
http://netpvrpa.cdn.elisaviihde.fi/stream.php?id=1812084&...
```

VLC playlist example
-----
```
$ cp examples/example_vlc.py .
$ python example_vlc.py -u username -f myplaylist.xspf
```

Creates XSPF playlist file (XML Shareable Playlist Format) containing recording stream links from first level folders.

http://en.wikipedia.org/wiki/XML_Shareable_Playlist_Format

![Build Status](https://raw.githubusercontent.com/enyone/elisaviihde/master/examples/example_playlist.png)


script.module.elisaviihde/changelog.txt
1.5
-----

1.4
-----
Video stream identifying fix
- Keyword for identifying line with video stream changed

Mark recording as watched
- Added ability to mark recording as watched

1.3
-----
Changed API endpoint URI to elisaviihde.fi (no more beta.elisaviihde.fi)

1.2
-----
Added KODI addon packaging
- Added script to create KODI/XBMC compatible zip addon package file.

1.1
-----
Deep folder structures and full list of recordings per folder
- Added support for deep folder structures and retrieves now a full list of recordings per folder

1.0
-----
First release using beta.elisaviihde.fi
- Supported functions: login, logout, folders, recordings, program, streamuri



tests.py
# Elisa Viihde API Python implementation tests
# License: GPLv3
# For-version: 1.5
# Author: enyone

import elisaviihde
from nose.tools import *
from httmock import urlmatch, HTTMock

# Mocks
@urlmatch(netloc=r'(.*\.)?elisaviihde\.fi$')
def elisaviihde_api_mock(url, request):
  if url.path == "/":
    return {'status_code': 200, 'content': '<html></html>'}
  elif url.path == "/api/sso/authcode":
    return {'status_code': 200, 'content': '{"code":"dummy-token"}'}
  elif url.path == "/api/user":
    return {'status_code': 200, 'content': '{"username":"dummy-user"}'}
  elif url.path == "/api/user/logout":
    return {'status_code': 200, 'content': '{}'}
  elif url.path == "/tallenteet/api/folders":
    return {'status_code': 200, 'content': '{"folders":[{"id":0,"folders":[{"id":1,"name":"dummy-folder"}]}]}'}
  elif url.path == "/tallenteet/api/folder/0":
    return {'status_code': 200, 'content': '{"recordingsCount":1}'}
  elif url.path == "/tallenteet/api/recordings/0":
    return {'status_code': 200, 'content': '[{"name":"dummy-recording"}]'}
  elif url.path == "/tallenteet/api/watched/0":
    return {'status_code': 200, 'content': '{}'}
  elif url.path == "/tallenteet/katso/0":
    return {'status_code': 200, 'content': 'data-section="recording-player" data-url="http://test.com/test"'}
  elif url.path == "/ohjelmaopas/ohjelma/1234":
    return {'status_code': 200, 'content': '\n<p itemprop="name">dummy-channel-name</p>\n'
              + '<p itemprop="description">dummy-service-description</p>\n'
              + '<span itemprop="startDate">01.02.2014 13:14</span>\n'
              + '<h3 itemprop="name" id="data-programid">dummy-service-name</h3>\n'}
  elif url.path == "/ohjelmaopas/ohjelma/1239":
    return {'status_code': 200, 'content': '\n<p itemprop="name">dummy-channel-name</p>\n'
              + '<p itemprop="description">dummy-service-description</p>\n'
              + '<span itemprop="startDate">01.02.2014a 13:14</span>\n'
              + '<h3 itemprop="name" id="data-programid">dummy-service-name</h3>\n'}
  else:
    return {'status_code': 500}

@urlmatch(netloc=r'(.*\.)?elisaviihde\.fi$')
def elisaviihde_api_mock_asshole(url, request):
  return {'status_code': 500}

@urlmatch(netloc=r'(.*\.)?elisaviihde\.fi$')
def elisaviihde_api_mock_badjson(url, request):
  return {'status_code': 200, 'content': '{"rew. ""ssdfg}  s'}

@urlmatch(netloc=r'(.*\.)?elisa\.fi$')
def elisaviihde_sso_mock(url, request):
  if url.path == "/sso/login":
    return {'status_code': 200}
  else:
    return {'status_code': 500}

@urlmatch(netloc=r'(.*\.)?elisa\.fi$')
def elisaviihde_sso_mock_asshole(url, request):
  return {'status_code': 500}

# Tests
def test_elisa_init_ok():
  with HTTMock(elisaviihde_api_mock):
    elisa = elisaviihde.elisaviihde(False)
  assert elisa.islogged() == False

@raises(Exception)
def test_elisa_init_fail():
  with HTTMock(elisaviihde_api_mock_asshole):
    elisa = elisaviihde.elisaviihde(False)

def test_elisa_login_ok():
  with HTTMock(elisaviihde_api_mock, elisaviihde_sso_mock):
    elisa = elisaviihde.elisaviihde(False)
    elisa.login("foo", "bar")
    assert elisa.islogged() == True
  assert elisa.gettoken() == "dummy-token"

@raises(Exception)
def test_elisa_login_fail():
  with HTTMock(elisaviihde_api_mock, elisaviihde_sso_mock_asshole):
    elisa = elisaviihde.elisaviihde(False)
    elisa.login("foo", "bar")

@raises(Exception)
def test_elisa_login_fail2():
  with HTTMock(elisaviihde_api_mock_badjson, elisaviihde_sso_mock):
    elisa = elisaviihde.elisaviihde(False)
    elisa.login("foo", "bar")

def test_elisa_logout_ok():
  with HTTMock(elisaviihde_api_mock, elisaviihde_sso_mock):
    elisa = elisaviihde.elisaviihde(False)
    elisa.login("foo", "bar")
    elisa.close()
  assert elisa.gettoken() == None

def test_elisa_user_ok():
  with HTTMock(elisaviihde_api_mock, elisaviihde_sso_mock):
    elisa = elisaviihde.elisaviihde(False)
    elisa.login("foo", "bar")
    user = elisa.getuser()
  assert user["username"] == "dummy-user"

@raises(Exception)
def test_elisa_user_fail():
  with HTTMock(elisaviihde_api_mock_badjson, elisaviihde_sso_mock):
    elisa = elisaviihde.elisaviihde(False)
    elisa.login("foo", "bar")
    user = elisa.getuser()

def test_elisa_sessions():
  with HTTMock(elisaviihde_api_mock, elisaviihde_sso_mock):
    elisa = elisaviihde.elisaviihde(False)
    elisa.login("foo", "bar")
    elisa.setsession({"foo":"123"})
    assert elisa.getsession() == {"foo":"123"}

def test_elisa_folders():
  with HTTMock(elisaviihde_api_mock, elisaviihde_sso_mock):
    elisa = elisaviihde.elisaviihde(False)
    elisa.login("foo", "bar")
    folders = elisa.getfolders()
  assert folders[0]["name"] == "dummy-folder"

def test_elisa_recordings():
  with HTTMock(elisaviihde_api_mock, elisaviihde_sso_mock):
    elisa = elisaviihde.elisaviihde(False)
    elisa.login("foo", "bar")
    recordings = elisa.getrecordings(0)
  assert recordings[0]["name"] == "dummy-recording"

def test_elisa_program():
  with HTTMock(elisaviihde_api_mock, elisaviihde_sso_mock):
    elisa = elisaviihde.elisaviihde(False)
    elisa.login("foo", "bar")
    program = elisa.getprogram(1234)
  assert program["serviceName"] == "dummy-channel-name"
  assert program["name"] == "dummy-service-name"
  assert program["startTimeUTC"] > 1300000000000
  assert program["description"] == "dummy-service-description"

def test_elisa_program_fail():
  with HTTMock(elisaviihde_api_mock, elisaviihde_sso_mock):
    elisa = elisaviihde.elisaviihde(False)
    elisa.login("foo", "bar")
    program = elisa.getprogram(1239)
  assert program["serviceName"] == "dummy-channel-name"
  assert program["description"] == "dummy-service-description"

def test_elisa_streamuri():
  with HTTMock(elisaviihde_api_mock, elisaviihde_sso_mock):
    elisa = elisaviihde.elisaviihde(False)
    elisa.login("foo", "bar")
    streamuri = elisa.getstreamuri(0)
  assert streamuri == "http://test.com/test"

def test_elisa_watched_ok():
  with HTTMock(elisaviihde_api_mock, elisaviihde_sso_mock):
    elisa = elisaviihde.elisaviihde(False)
    elisa.login("foo", "bar")
    elisa.markwatched(0)

@raises(Exception)
def test_elisa_watched_fail():
  with HTTMock(elisaviihde_api_mock_badjson, elisaviihde_sso_mock):
    elisa = elisaviihde.elisaviihde(False)
    elisa.login("foo", "bar")
    elisa.markwatched(0)


examples/example.py
# Elisa Viihde API Python implementation usage demo

import getopt, sys, getpass, elisaviihde

def main():
  # Parse command line args
  if len(sys.argv[1:]) < 2:
    print "ERROR: Usage:", sys.argv[0], "[-u username]" 
    sys.exit(2)
  try:
    opts, args = getopt.getopt(sys.argv[1:], "u:v", ["username"])
  except getopt.GetoptError as err:
    print "ERROR:", str(err)
    sys.exit(2) 
  
  # Init args
  username = ""
  verbose = False
  
  # Read arg data
  for o, a in opts:
    if o == "-v":
      verbose = True
    elif o in ("-u", "--username"):
      username = a
    else:
      assert False, "unhandled option"
  
  # Ask password securely on command line
  password = getpass.getpass('Password: ')
  
  # Init elisa session
  try:
    elisa = elisaviihde.elisaviihde(verbose)
  except Exception as exp:
    print "ERROR: Could not create elisa session"
    sys.exit(1)
  
  # Login
  try:
    elisa.login(username, password)
  except Exception as exp:
    print "ERROR: Login failed, check username and password"
    sys.exit(1)
  
  # Read and print recording folders
  folders = elisa.getfolders()
  
  print "\nFound folders:"
  for folder in folders:
    print str(folder["id"]) + ": " + folder["name"]
  
  # Read and print recording folders
  folderid = folders[0]["id"]
  recordings = elisa.getrecordings(folderid)
  
  print "\nFound recordings from folder " + str(folderid) + ":"
  for recording in recordings:
    print str(recording["programId"]) + ": " + recording["name"] + " (" + recording["startTimeFormatted"] + ")"
  
  # Get recording stream uri from first recording
  programid = recordings[0]["programId"]
  streamuri = elisa.getstreamuri(programid)
  
  print "\nFound stream uri from recording " + str(programid) + ":"
  print streamuri
  
  # Close session
  elisa.close()

if __name__ == "__main__":
  main()



examples/example_vlc.py
# Elisa Viihde API Python implementation usage demo
# Creates vlc xspf playlist file containing all recordings
# (first page only, 10 per page) from all folders recursively

import getopt, sys, cgi, datetime, getpass, elisaviihde
import xml.etree.ElementTree as et

def getxmlheader():
  return """<?xml version="1.0" encoding="UTF-8"?>
            <playlist xmlns="http://xspf.org/ns/0/" xmlns:vlc="http://www.videolan.org/vlc/playlist/ns/0/" version="1">
            <title>Elisa Viihde</title><trackList>
         """

def getxmlfooter():
  return """</trackList><extension application="http://www.videolan.org/vlc/playlist/0"><vlc:item tid="0"/>
            </extension></playlist>
         """

def getxmltrack(folder, recording, uri):
  return """<track><location>%s</location><title>%s</title><creator>%s</creator>
            <album>%s</album><annotation>%s</annotation><duration>%d</duration>
            <image>%s</image>
            <extension application="http://www.videolan.org/vlc/playlist/0"><vlc:id>%d</vlc:id>
            <vlc:option>network-caching=1000</vlc:option></extension></track>
         """ % (cgi.escape(uri),
                cgi.escape(recording["name"]),
                cgi.escape(datetime.datetime.fromtimestamp(recording["startTimeUTC"]/1000).strftime('%Y-%m-%d %H:%M:%S')
                             + " " + recording["serviceName"]),
                cgi.escape(folder["name"]) if folder and "name" in folder else "",
                cgi.escape(recording["description"]) if "description" in recording else "",
                ((recording["endTimeUTC"]) - (recording["startTimeUTC"])),
                cgi.escape(recording["thumbnail"]) if "thumbnail" in recording else "",
                recording["programId"])

def main():
  # Parse command line args
  if len(sys.argv[1:]) < 2:
    print "ERROR: Usage:", sys.argv[0], "[-u username] [-f outputfile]" 
    sys.exit(2)
  try:
    opts, args = getopt.getopt(sys.argv[1:], "u:f:v", ["username", "outputfile"])
  except getopt.GetoptError as err:
    print "ERROR:", str(err)
    sys.exit(2)
  
  # Init args
  username = ""
  outputfile = "playlist.xspf"
  verbose = False
  
  # Read arg data
  for o, a in opts:
    if o == "-v":
      verbose = True
    elif o in ("-u", "--username"):
      username = a
    elif o in ("-f", "--outputfile"):
      outputfile = a
    else:
      assert False, "unhandled option"
  
  # Ask password securely on command line
  password = getpass.getpass('Password: ')
  
  # Init elisa session
  try:
    elisa = elisaviihde.elisaviihde(verbose)
  except Exception as exp:
    print "ERROR: Could not create elisa session"
    sys.exit(1)
  
  # Login
  try:
    elisa.login(username, password)
  except Exception as exp:
    print "ERROR: Login failed, check username and password"
    sys.exit(1)
  
  filehandle = open(outputfile, 'w')
  filehandle.write(getxmlheader())
  
  # Read top directory and persist to a file
  recordings = elisa.getrecordings(0, 0)
  for recording in recordings:
    streamuri = elisa.getstreamuri(recording["programId"])
    filehandle.write(getxmltrack(None, recording, streamuri).encode('utf8'))

  # Walk sub-directories recursively and persist to a file
  folders = elisa.getfolders()
  for folder in folders:
    recordings = elisa.getrecordings(folder["id"], 0)
    for recording in recordings:
      streamuri = elisa.getstreamuri(recording["programId"])
      filehandle.write(getxmltrack(folder, recording, streamuri).encode('utf8'))
  
  filehandle.write(getxmlfooter())
  filehandle.close()
  
  # Close session
  elisa.close()

if __name__ == "__main__":
  main()



elisaviihde.py
# Elisa Viihde API Python implementation
# License: GPLv3
# Author: enyone
# Version: 1.5
# https://github.com/enyone/elisaviihde

import requests, json, re, time, datetime, math

class elisaviihde:
  # Init args
  verbose = False
  baseurl = "https://elisaviihde.fi"
  ssobaseurl = "https://id.elisa.fi"
  session = None
  authcode = None
  userinfo = None
  inited = False
  verifycerts = False
  
  def __init__(self, verbose=False):
    # Init session to store cookies
    self.verbose = verbose
    self.session = requests.Session()
    self.session.headers.update({"Referer": self.baseurl + "/"})
    
    # Make initial request to get session cookie
    if self.verbose: print "Initing session..."
    
    init = self.session.get(self.baseurl + "/", verify=self.verifycerts)
    self.checkrequest(init.status_code)
  
  def login(self, username, password):
    # Get sso auth token
    if self.verbose: print "Getting single-sign-on token..."
    token = self.session.post(self.baseurl + "/api/sso/authcode",
                              data={"username": username},
                              headers={"Content-type": "application/x-www-form-urlencoded; charset=UTF-8",
                                       "X-Requested-With": "XMLHttpRequest"},
                              verify=self.verifycerts)
    self.checkrequest(token.status_code)
    try:
      self.authcode = token.json()["code"]
    except ValueError as err:
      raise Exception("Could not fetch sso token", err)
    
    # Login with token
    if self.verbose: print "Logging in with single-sign-on token..."
    login = self.session.post(self.ssobaseurl + "/sso/login",
                              data=json.dumps({"accountId": username,
                                               "password": password,
                                               "authCode": self.authcode,
                                               "suppressErrors": True}),
                              headers={"Content-type": "application/json; charset=UTF-8",
                                       "Origin": self.baseurl},
                              verify=self.verifycerts)
    self.checkrequest(login.status_code)
    
    # Login with username and password
    if self.verbose: print "Logging in with username and password..."
    user = self.session.post(self.baseurl + "/api/user",
                             data={"username": username,
                                   "password": password},
                             headers={"Content-type": "application/x-www-form-urlencoded; charset=UTF-8",
                                      "X-Requested-With": "XMLHttpRequest"},
                             verify=self.verifycerts)
    self.checkrequest(user.status_code)
    try:
      self.userinfo = user.json()
    except ValueError as err:
      raise Exception("Could not fetch user information", err)
    self.inited = True
  
  def islogged(self):
    if self.inited:
      return True
    try:
      logincheck = self.session.get(self.baseurl + "/tallenteet/api/folders",
                                    headers={"X-Requested-With": "XMLHttpRequest"},
                                    verify=self.verifycerts)
      self.checkrequest(logincheck.status_code)
      self.inited = True
      return True
    except Exception as err:
      return False
    
  def checklogged(self):
    if not self.islogged():
      raise Exception("Not logged in")
  
  def checkrequest(self, statuscode):
    if not statuscode == requests.codes.ok:
      raise Exception("API request failed with error code: " + str(statuscode))
  
  def close(self):
    if self.verbose: print "Logging out and closing session..."
    logout = self.session.post(self.baseurl + "/api/user/logout",
                               headers={"X-Requested-With": "XMLHttpRequest"},
                               verify=self.verifycerts)
    self.session.close()
    self.userinfo = None
    self.authcode = None
    self.inited = False
    self.checkrequest(logout.status_code)
  
  def gettoken(self):
    return self.authcode
  
  def getuser(self):
    return self.userinfo
  
  def getsession(self):
    return requests.utils.dict_from_cookiejar(self.session.cookies)
    
  def setsession(self, cookies):  
    self.session.cookies=requests.utils.cookiejar_from_dict(cookies)
  
  def walk(self, tree, parent=None):
    # Walk folder tree recursively
    flat = []
    subtree = None
    if "folders" in tree:
      if len(tree["folders"]) > 0:
        subtree = tree["folders"]
      del tree["folders"]
    if "id" in tree:
      tree["parentFolder"] = parent if tree["id"] > 0 else None
      flat.append(tree)
    if subtree:
      for folder in subtree:
        flat += self.walk(folder, (tree["id"] if "id" in tree else 0))
    return flat
  
  def getfolders(self, folderid=0):
    # Get folders
    if self.verbose: print "Getting folders..."
    self.checklogged()
    folders = self.session.get(self.baseurl + "/tallenteet/api/folders",
                               headers={"X-Requested-With": "XMLHttpRequest"},
                               verify=self.verifycerts)
    self.checkrequest(folders.status_code)
    return [folder for folder in self.walk(folders.json()) if folder["parentFolder"] == folderid]
  
  def getfolderstatus(self, folderid=0):
    # Get folder info
    if self.verbose: print "Getting folder info..."
    self.checklogged()
    folder = self.session.get(self.baseurl + "/tallenteet/api/folder/" + str(folderid),
                               headers={"X-Requested-With": "XMLHttpRequest"},
                               verify=self.verifycerts)
    self.checkrequest(folder.status_code)
    return folder.json()
  
  def getrecordings(self, folderid=0, page=None, sortby="startTime", sortorder="desc", status="all"):
    # Get recordings from folder
    self.checklogged()
    if self.verbose: print "Getting recordings..."
    recordings = []
    if page == None:
      folder = self.getfolderstatus(folderid)
      # Append rest of pages to list (50 recordings per page)
      maxpage = int(math.floor(folder["recordingsCount"] / 50))
      if maxpage > 0:
        pages = range(0, maxpage)
      else:
        pages = [0]
    else:
      pages = [page]
    for pageno in pages:
      recordingspp = self.session.get(self.baseurl + "/tallenteet/api/recordings/" + str(folderid)
                                        + "?page=" + str(pageno)
                                        + "&sortBy=" + str(sortby)
                                        + "&sortOrder=" + str(sortorder)
                                        + "&watchedStatus=" + str(status),
                                      headers={"X-Requested-With": "XMLHttpRequest"},
                                      verify=self.verifycerts)
      self.checkrequest(recordingspp.status_code)
      recordings += recordingspp.json()
    return recordings
  
  def getprogram(self, programid=0):
    # Parse program information
    self.checklogged()
    if self.verbose: print "Getting program info..."
    uridata = self.session.get(self.baseurl + "/ohjelmaopas/ohjelma/" + str(programid), verify=self.verifycerts)
    self.checkrequest(uridata.status_code)
    programname = ""
    programdesc = ""
    programsrvc = ""
    programtime = 0
    try:
      for line in uridata.text.split("\n"):
        if "itemprop=\"name\"" in line and "data-programid" in line:
          programname = re.findall('<h3.*?>(.*?)</h3>', line)[0]
        elif "itemprop=\"description\"" in line:
          programdesc = re.findall('<p.*?>(.*?)</p>', line)[0]
        elif "itemprop=\"name\"" in line:
          programsrvc = re.findall('<p.*?>(.*?)</p>', line)[0]
        elif "itemprop=\"startDate\"" in line:
          programtimestr = re.findall('<span.*?>(.*?)</span>', line)[0]
          programtime = int(datetime.datetime.fromtimestamp(
                              time.mktime(time.strptime(programtimestr,
                                                        "%d.%m.%Y %H:%M"))).strftime("%s"))
          programtime = programtime * 1000
    except Exception as exp:
      print "ERROR:", str(exp)
    except Error as exp:
      print "ERROR:", str(exp)
    
    return {"name": programname, "description": programdesc, "serviceName": programsrvc, "startTimeUTC": programtime}
  
  def getstreamuri(self, programid=0):
    # Parse recording stream uri for program
    self.checklogged()
    if self.verbose: print "Getting stream uri info..."
    uridata = self.session.get(self.baseurl + "/tallenteet/katso/" + str(programid), verify=self.verifycerts)
    self.checkrequest(uridata.status_code)
    for line in uridata.text.split("\n"):
      if "recording-player" in line:
        return re.findall('http[s]?://(?:[a-zA-Z]|[0-9]|[$-_@.&+]|[!*\(\),]|(?:%[0-9a-fA-F][0-9a-fA-F]))+', line)[0]

  def markwatched(self, programid=0):
    # Mark recording as watched
    if self.verbose: print "Marking as watched..."
    self.checklogged()
    watched = self.session.get(self.baseurl + "/tallenteet/api/watched/" + str(programid),
                               headers={"X-Requested-With": "XMLHttpRequest"},
                               verify=self.verifycerts)
    self.checkrequest(watched.status_code)

